BlueprintUI/Sources/AttributedText/AttributedText.swift

```diff
@@ -0,0 +1,158 @@
+        newString.append(rhs.mutableAttributedString)
+        return AttributedText(newString)
+    }
+
+    private var entireRange: Range<String.Index> {
+        string.startIndex..<string.endIndex
+    }
+
+    private mutating func addAttributes(attributes: TextAttributeContainer, to range: NSRange) {
+        if !isKnownUniquelyReferenced(&mutableAttributedString) {
+            mutableAttributedString = NSMutableAttributedString(attributedString: mutableAttributedString)
+        }
+
+        mutableAttributedString.addAttributes(attributes.storage, range: range)
+    }
+
+    /// The implementation of this function may not be intuitive, but is necessary due to how enumerating attributes
+    /// works. For example, consider the string "Block" with the font attribute set for the whole string, the color
+    /// red set for the range of "Blo" and the color blue set for the range of "ck".
+    ///
+    /// If we get the attributes for the entire range of the string, the desired output is a
+    /// `TextAttributeContainer` with only the `font` specified, since it's the only attribute that applies to that
+    /// whole range. If we enumerate the attributes of the string however, the system will give us two results:
+    ///
+    /// - The range of "Blo" with the font and red color in the dictionary
+    /// - The range of "ck" with the font and the blue color in the dictionary
+    ///
+    /// Now we need to "merge" attributes that match across the whole range, but in practice we cannot do that: the
+    /// dictionary contains values of type "any" so we cannot compare them. We know that a font attribute is present
+    /// in both the first range and second range, but we don't know if the value is equal, and therefor valid for
+    /// the entire range.
+    ///
+    /// This leads the implementation below: first, collect the attributes present in the range at all. Secondly,
+    /// enumerate the attributes individually. The system finds the longest effective range of attributes for us, so
+    /// if the range of the found attribute matches the range we're checking, add the attribute to our text container.
+    ///
+    private func makeAttributeStore(range: NSRange) -> TextAttributeContainer {
+        var store = TextAttributeContainer.empty
+        var attributesInRange: Set<NSAttributedString.Key> = []
+
+        mutableAttributedString.enumerateAttributes(
+            in: range,
+            options: []
+        ) { attributes, _, _ in
+            attributesInRange.formUnion(attributes.keys)
+        }
+
+        for attribute in attributesInRange {
+            mutableAttributedString.enumerateAttribute(
+                attribute,
+                in: range,
+                options: []
+            ) { value, attributeRange, _ in
+                if let value = value, attributeRange == range {
+                    store.storage[attribute] = value
+                }
+            }
```

Comment:
So if I'm following what you're saying here:

> If the attributes extend beyond the range they'll still be returned! It's only attributes that apply to part of the range that aren't returned

The `attributedRange` passed into this block by `enumerateAttribute` will be no larger than `range`, even if that attributes's range does extend outside of `range`?

Reply:
Yes! Although it's not documented well, that's the behavior I'm seeing (and expected, tbh)